Bug: Stepless subqueries fail

This query fails:
`avg_over_time(up[1h:])`

it results in a panic:

```
panic: runtime error: invalid memory address or nil pointer dereference

goroutine 588010 [running]:

runtime/debug.Stack()
    /usr/local/go/src/runtime/debug/stack.go:26 +0x64

github.com/metrico/qryn/v4/reader/controller.tamePanic({0x19658e0, 0x150f45f4bbc0}, 0x150f457cd680)
    /src/reader/controller/utils.go:78 +0x3c

panic({0x12739c0?, 0x2a17a60?})
    /usr/local/go/src/runtime/panic.go:860 +0x12c

github.com/prometheus/prometheus/promql.(*Engine).getLastSubqueryInterval(0x150f41f02600, {0x150f46b02300?, 0x6, 0x1940ea8?})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/engine.go:1022 +0x70

github.com/prometheus/prometheus/promql.(*Engine).populateSeries.func1({0x1966d90?, 0x150f47cf2f00}, {0x150f46b02300, 0x6, 0x8})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/engine.go:1039 +0xe0

github.com/prometheus/prometheus/promql/parser.inspector.Visit(0x150f47201500, {0x1966d90?, 0x150f47cf2f00?}, {0x150f46b02300?, 0x1161780?, 0x150f48df8db0?})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:385 +0x48

github.com/prometheus/prometheus/promql/parser.Walk({0x19578a0?, 0x150f47201500?}, {0x1966d90, 0x150f47cf2f00}, {0x150f46b02300, 0x6, 0x8})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:352 +0x5c

github.com/prometheus/prometheus/promql/parser.Walk-range1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:361

github.com/prometheus/prometheus/promql/parser.Walk.ChildrenIter.func1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:416

github.com/prometheus/prometheus/promql/parser.Walk({0x19578a0?, 0x150f47201500?}, {0x1966ee0, 0x150f45f4bda0}, {0x150f46b02300, 0x5, 0x8})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:357 +0x11d8

github.com/prometheus/prometheus/promql/parser.Walk-range1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:361

github.com/prometheus/prometheus/promql/parser.Walk.ChildrenIter.func1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:437

github.com/prometheus/prometheus/promql/parser.Walk({0x19578a0?, 0x150f47201500?}, {0x1967590, 0x150f471aeb00}, {0x150f47201540, 0x4, 0x4})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:357 +0x954

github.com/prometheus/prometheus/promql/parser.Walk-range1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:361

github.com/prometheus/prometheus/promql/parser.Walk.ChildrenIter.func1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:435

github.com/prometheus/prometheus/promql/parser.Walk({0x19578a0?, 0x150f47201500?}, {0x1966df0, 0x150f45f4be00}, {0x150f47201540, 0x3, 0x4})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:357 +0x72c

github.com/prometheus/prometheus/promql/parser.Walk-range1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:361

github.com/prometheus/prometheus/promql/parser.Walk.ChildrenIter.func1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:430

github.com/prometheus/prometheus/promql/parser.Walk({0x19578a0?, 0x150f47201500?}, {0x1966e20, 0x150f4840dad0}, {0x150f47201540, 0x2, 0x4})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:357 +0xb54

github.com/prometheus/prometheus/promql/parser.Walk-range1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:361

github.com/prometheus/prometheus/promql/parser.Walk.ChildrenIter.func1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:427

github.com/prometheus/prometheus/promql/parser.Walk({0x19578a0?, 0x150f47201500?}, {0x1966e50, 0x150f47200f40}, {0x150f47201540, 0x1, 0x4})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:357 +0xe08

github.com/prometheus/prometheus/promql/parser.Walk-range1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:361

github.com/prometheus/prometheus/promql/parser.Walk.ChildrenIter.func1(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:424

github.com/prometheus/prometheus/promql/parser.Walk({0x19578a0?, 0x150f47201500?}, {0x1966e50, 0x150f47201280}, {0x150f47201540, 0x0, 0x4})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:357 +0xcec

github.com/prometheus/prometheus/promql/parser.Inspect(...)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/parser/ast.go:398

github.com/prometheus/prometheus/promql.(*Engine).populateSeries(0x150f41f02600, {0x196a808, 0x150f4840dcb0}, {0x196c3e0, 0x150f471aeba0}, 0x150f41d9ee10)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/engine.go:1035 +0x11c

github.com/prometheus/prometheus/promql.(*Engine).execEvalStmt(0x150f41f02600, {0x196a808, 0x150f4840dc50}, 0x150f47ae57a0, 0x150f41d9ee10)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/engine.go:783 +0x148

github.com/prometheus/prometheus/promql.(*Engine).exec(0x150f41f02600, {0x196a808, 0x150f4840dc50}, 0x150f47ae57a0)
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/engine.go:744 +0x268

github.com/prometheus/prometheus/promql.(*query).Exec(0x150f47ae57a0, {0x196a808, 0x150f4840cdb0})
    /go/pkg/mod/github.com/prometheus/prometheus@v0.311.3/promql/engine.go:254 +0x14c

github.com/metrico/qryn/v4/reader/controller.(*PromQueryRangeController).QueryInstant(0x150f4200ac78, {0x19658e0, 0x150f45f4bbc0}, 0x150f457cd680)
    /src/reader/controller/prom_query_instant.go:54 +0x29c

net/http.HandlerFunc.ServeHTTP(0x19658b0?, {0x19658e0?, 0x150f45f4bbc0?}, 0x0?)
    /usr/local/go/src/net/http/server.go:2286 +0x38

github.com/metrico/qryn/v4/reader/utils/middleware.AcceptEncodingMiddleware.func1({0x19658b0, 0x150f47d53c80}, 0x150f457cd680)
    /src/reader/utils/middleware/accept_encoding.go:19 +0xb4

net/http.HandlerFunc.ServeHTTP(0xede579fc0bf0?, {0x19658b0?, 0x150f47d53c80?}, 0x150f457cd680?)
    /usr/local/go/src/net/http/server.go:2286 +0x38

github.com/metrico/qryn/v4/reader/utils/middleware.LoggingMiddleware.func1.1({0x1965bc0, 0x150f456ac690}, 0x150f457cd680)
    /src/reader/utils/middleware/logging.go:25 +0xb4

net/http.HandlerFunc.ServeHTTP(0x150f457cd540?, {0x1965bc0?, 0x150f456ac690?}, 0x4?)
    /usr/local/go/src/net/http/server.go:2286 +0x38

github.com/gorilla/mux.(*Router).ServeHTTP(0x150f41d120c0, {0x1965bc0, 0x150f456ac690}, 0x150f457cd400)
    /go/pkg/mod/github.com/gorilla/mux@v1.8.1/mux.go:212 +0x188

net/http.serverHandler.ServeHTTP({0x195f718?}, {0x1965bc0?, 0x150f456ac690?}, 0x6?)
    /usr/local/go/src/net/http/server.go:3311 +0xb0

net/http.(*conn).serve(0x150f465bcbd0, {0x196a808, 0x150f41c24270})
    /usr/local/go/src/net/http/server.go:2073 +0x51c

created by net/http.(*Server).Serve in goroutine 1
    /usr/local/go/src/net/http/server.go:3464 +0x37c
```

--- a/reader/router/prometheus_query_range.go
+++ b/reader/router/prometheus_query_range.go
@@ -15,23 +15,38 @@ import (
 	api_v1 "github.com/prometheus/prometheus/web/api/v1"
 )
 
-func RoutePrometheusQueryRange(app *mux.Router, dataSession model.IDBRegistry,
-	stats bool,
-) {
+// defaultSubqueryInterval is used as the resolution step for subqueries that
+// omit one (e.g. `up[1h:]`). It matches Prometheus' default evaluation interval.
+const defaultSubqueryInterval = time.Minute
+
+// NewPromEngine builds the PromQL engine used to serve /api/v1/query and
+// /api/v1/query_range requests.
+func NewPromEngine(maxSamples int) *promql.Engine {
 	slogLogger := slog.New(slog.NewJSONHandler(logger.Logger.Out, &slog.HandlerOptions{
 		Level: slog.LevelDebug,
 	}))
-	eng := promql.NewEngine(promql.EngineOpts{
-		Logger:                   slogLogger,
-		Reg:                      nil,
-		MaxSamples:               config.Cloki.Setting.SYSTEM_SETTINGS.MetricsMaxSamples,
-		Timeout:                  time.Second * 30,
-		ActiveQueryTracker:       nil,
-		LookbackDelta:            0,
-		NoStepSubqueryIntervalFn: nil,
-		EnableAtModifier:         false,
-		EnableNegativeOffset:     false,
+	return promql.NewEngine(promql.EngineOpts{
+		Logger:             slogLogger,
+		Reg:                nil,
+		MaxSamples:         maxSamples,
+		Timeout:            time.Second * 30,
+		ActiveQueryTracker: nil,
+		LookbackDelta:      0,
+		// A non-nil function is required: the engine calls it for subqueries
+		// that omit a resolution step (e.g. `up[1h:]`). Leaving it nil panics
+		// with a nil pointer dereference in getLastSubqueryInterval.
+		NoStepSubqueryIntervalFn: func(int64) int64 {
+			return defaultSubqueryInterval.Milliseconds()
+		},
+		EnableAtModifier:     false,
+		EnableNegativeOffset: false,
 	})
+}
+
+func RoutePrometheusQueryRange(app *mux.Router, dataSession model.IDBRegistry,
+	stats bool,
+) {
+	eng := NewPromEngine(config.Cloki.Setting.SYSTEM_SETTINGS.MetricsMaxSamples)
 	svc := service.CLokiQueriable{
 		ServiceData: model.ServiceData{Session: dataSession},
 	}

--- a/reader/router/prometheus_query_range_test.go
+++ b/reader/router/prometheus_query_range_test.go
@@ -1,0 +1,54 @@
+package router
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/util/annotations"
+)
+
+// emptyQueryable is a stub storage.Queryable that returns no series. It lets us
+// drive the PromQL engine through query planning/execution without a backend.
+type emptyQueryable struct{}
+
+func (emptyQueryable) Querier(int64, int64) (storage.Querier, error) {
+	return emptyQuerier{}, nil
+}
+
+type emptyQuerier struct{}
+
+func (emptyQuerier) Select(context.Context, bool, *storage.SelectHints, ...*labels.Matcher) storage.SeriesSet {
+	return storage.EmptySeriesSet()
+}
+
+func (emptyQuerier) LabelValues(context.Context, string, *storage.LabelHints, ...*labels.Matcher) ([]string, annotations.Annotations, error) {
+	return nil, nil, nil
+}
+
+func (emptyQuerier) LabelNames(context.Context, *storage.LabelHints, ...*labels.Matcher) ([]string, annotations.Annotations, error) {
+	return nil, nil, nil
+}
+
+func (emptyQuerier) Close() error { return nil }
+
+// TestSubqueryWithoutStep guards against a regression where a subquery that
+// omits its resolution step (e.g. `up[1h:]`) panicked with a nil pointer
+// dereference because the engine's NoStepSubqueryIntervalFn was nil.
+func TestSubqueryWithoutStep(t *testing.T) {
+	eng := NewPromEngine(5000000)
+	ctx := context.Background()
+
+	q, err := eng.NewInstantQuery(ctx, emptyQueryable{}, nil, "avg_over_time(up[1h:])", time.Unix(0, 0))
+	if err != nil {
+		t.Fatalf("NewInstantQuery: %v", err)
+	}
+	defer q.Close()
+
+	res := q.Exec(ctx)
+	if res.Err != nil {
+		t.Fatalf("Exec: %v", res.Err)
+	}
+}

